hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,712 @@
+            "publish_type": publish_type,
+        }
+
+    def _get_template(self, template_name, settings):
+        """Return the configured template for the supplied name.
+
+        TODO: should publish2 populate "settings" with the configured template?
+              currently it just stores the name.
+        """
+
+        publisher = self.parent
+        template_name = settings[template_name].value
+        return publisher.get_template_by_name(template_name)
+
+
+def _maya_find_additional_session_dependencies():
+    """
+    Find additional dependencies from the session
+    """
+
+    # default implementation looks for references and
+    # textures (file nodes) and returns any paths that
+    # match a template defined in the configuration
+    ref_paths = set()
+
+    # first let's look at maya references
+    ref_nodes = cmds.ls(references=True)
+    for ref_node in ref_nodes:
+        # get the path:
+        ref_path = cmds.referenceQuery(ref_node, filename=True)
+        # make it platform dependent
+        # (maya uses C:/style/paths)
+        ref_path = ref_path.replace("/", os.path.sep)
+        if ref_path:
+            ref_paths.add(ref_path)
+
+    # now look at file texture nodes
+    for file_node in cmds.ls(l=True, type="file"):
+        # ensure this is actually part of this session and not referenced
+        if cmds.referenceQuery(file_node, isNodeReferenced=True):
+            # this is embedded in another reference, so don't include it in
+            # the breakdown
+            continue
+
+        # get path and make it platform dependent
+        # (maya uses C:/style/paths)
+        texture_path = cmds.getAttr(
+            "%s.fileTextureName" % file_node).replace("/", os.path.sep)
+        if texture_path:
+            ref_paths.add(texture_path)
+
+    return list(ref_paths)
+
+
+def _session_path():
+    """
+    Return the path to the current session
+    :return:
+    """
+    path = cmds.file(query=True, sn=True)
+
+    if isinstance(path, unicode):
+        path = path.encode("utf-8")
+
+    return path
+
+
+def _save_session(path):
+    """
+    Save the current session to the supplied path.
+    """
+
+    # Maya can choose the wrong file type so we should set it here
+    # explicitly based on the extension
+    maya_file_type = None
+    if path.lower().endswith(".ma"):
+        maya_file_type = "mayaAscii"
+    elif path.lower().endswith(".mb"):
+        maya_file_type = "mayaBinary"
+
+    cmds.file(rename=path)
+
+    # save the scene:
+    if maya_file_type:
+        cmds.file(save=True, force=True, type=maya_file_type)
+    else:
+        cmds.file(save=True, force=True)
+
+
+def _get_save_as_action():
+    """
+
+    Simple helper for returning a log action dict for saving the session
+    """
+
+    engine = sgtk.platform.current_engine()
+
+    # default save callback
+    callback = cmds.SaveScene
+
+    # if workfiles2 is configured, use that for file save
```

Comment:
Modified this to allow file save via workfiles2 if configured. reasonable?

Reply:
great!